app/src/main/java/io/appium/settings/RecorderService.java

```diff
@@ -0,0 +1,135 @@
+
+        int result = START_STICKY;
+        final String action = intent != null ? intent.getAction() : null;
+        Log.v(TAG, "onStartCommand:action string: " + action);
+        if (ACTION_RECORDING_START.equals(action)) {
+            showNotification(); // TODO is this really necessary
+
+            MediaProjectionManager mMediaProjectionManager =
+                    (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
+
+            if (mMediaProjectionManager != null) {
+                startScreenRecord(mMediaProjectionManager, intent);
+            } else {
+                Log.e(TAG, "mMediaProjectionManager null");
+            }
+        } else if (ACTION_RECORDING_STOP.equals(action)) {
+            Log.v(TAG, "onStartCommand:intent=" + "stop");
+            stopScreenRecord();
+            result = START_NOT_STICKY;
+        }
+
+        return result;
+    }
+
+    /**
+     * start screen recording
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void startScreenRecord(MediaProjectionManager mediaProjectionManager,
+                                   final Intent intent) {
+        if (recorderThread == null) {
+            Log.v(TAG, "recorder thread null");
+
+            final int resultCode = intent.getIntExtra(ACTION_RECORDING_RESULT_CODE, 0);
+            // get MediaProjection
+            final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                    intent);
+            if (projection != null) {
+                Log.v(TAG, "projection different from null");
+                final DisplayMetrics metrics = getResources().getDisplayMetrics();
+                int rawWidth = metrics.widthPixels;
+                int rawHeight = metrics.heightPixels;
+                boolean isRotated = false;
+                if (rawWidth > rawHeight) {
+                    // landscape mode
+                    isRotated = true;
+                    /* TODO we need to rotate frames that comes from virtual screen before writing to file via muxer,
+                    *  for handling landscape mode properly, rotateYuvImages somehow fast and reliable
+                    *  for now, just flip width and height to fake and force portrait mode instead of landscape
+                    * */
+                    rawWidth = metrics.heightPixels;
+                    rawHeight = metrics.widthPixels;
+                }
+                Log.v(TAG, String.format("startRecording:(%d,%d)(%d,%d)",
+                        metrics.widthPixels, metrics.heightPixels, rawWidth, rawHeight));
+                String outputFilePath = intent.getStringExtra(ACTION_RECORDING_FILENAME);
+                if (outputFilePath != null) {
+                    recorderThread = new RecorderThread(projection, outputFilePath,
+                            rawWidth, rawHeight, isRotated);
+                    recorderThread.startRecording();
+                } else {
```

Comment:
same as above. Too much of nesting, thus hard to read.

Reply:
done